File: Sources/windows/BitTherapyCLR/src/pets/pets_builder.cpp
#include "pets_builder.h"

#include <iostream>

#include "../capabilities/capabilities.h"

PetsBuilder::PetsBuilder(
    const SpeciesRepository* speciesRepo, 
    const SpritesRepository* spritesRepo,
    const double animationFps,
    const double baseSize,
    const double scaleMultiplier,
    const double speedMultiplier
) : 
    nextId(1000),
    speciesRepo(speciesRepo),
    spritesRepo(spritesRepo),
    animationFps(animationFps),
    baseSize(baseSize),
    scaleMultiplier(scaleMultiplier),
    speedMultiplier(speedMultiplier)
{}

std::optional<Entity*> PetsBuilder::build(const std::string& speciesId, const Rect& gameBounds) {
    auto speciesOpt = speciesRepo->species(speciesId);
    auto spritesOpt = spritesRepo->sprites(speciesId);

    if (!speciesOpt.has_value() || !spritesOpt.has_value()) {
        std::cerr << "Species `" << speciesId << "` not found!" << std::endl;
        return std::nullopt;
    }
    auto species = speciesOpt.value();
    auto sprites = spritesOpt.value();  

    auto frame = Rect(
        rand() % (int)gameBounds.w,
        1.0, 
        baseSize * species->scale * scaleMultiplier, 
        baseSize * species->scale * scaleMultiplier
    );

    Entity* entity = new Entity(++nextId, animationFps, speedMultiplier, species, sprites, frame);

    bool isWallCrawler = std::find(species->capabilities.begin(), species->capabilities.end(), "WallCrawler") != species->capabilities.end();

    entity->addCapability(std::make_shared<RandomAnimations>());
    entity->addCapability(std::make_shared<LinearMovement>());
    entity->addCapability(std::make_shared<Gravity>(gameBounds.h));

    if (isWallCrawler) {
        entity->addCapability(std::make_shared<WallCrawler>(gameBounds));
    } else {
        entity->addCapability(std::make_shared<BounceWhenLateralBoundIsHit>(0, gameBounds.w));
    }

    return entity;
}

File: Sources/windows/App/Sources/Pets/PetsBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace App
{
    public class PetsBuilder
    {
        private uint nextId;
        private readonly SpeciesRepository speciesRepo;
        private readonly SpritesRepository spritesRepo;
        private readonly double animationFps;
        private readonly double baseSize;
        private readonly double scaleMultiplier;
        private readonly double speedMultiplier;

        public PetsBuilder(
            SpeciesRepository speciesRepo,
            SpritesRepository spritesRepo,
            double animationFps,
            double baseSize,
            double scaleMultiplier,
            double speedMultiplier
        )
        {
            nextId = 1000;
            this.speciesRepo = speciesRepo;
            this.spritesRepo = spritesRepo;
            this.animationFps = animationFps;
            this.baseSize = baseSize;
            this.scaleMultiplier = scaleMultiplier;
            this.speedMultiplier = speedMultiplier;
        }

        public Entity? Build(string speciesId, Rect gameBounds)
        {
            var speciesOpt = speciesRepo.GetSpecies(speciesId);
            var spritesOpt = spritesRepo.GetSprites(speciesId);

            if (speciesOpt == null || spritesOpt == null)
            {
                Console.Error.WriteLine($"Species `{speciesId}` not found!");
                return null;
            }
            var species = speciesOpt;
            var sprites = spritesOpt;

            var frame = new Rect(
                new Random().Next((int)gameBounds.W),
                1.0,
                baseSize * species.Scale * scaleMultiplier,
                baseSize * species.Scale * scaleMultiplier
            );

            var entity = new Entity(++nextId, animationFps, speedMultiplier, species, sprites, frame);

            bool isWallCrawler = species.Capabilities.Contains("WallCrawler");

            entity.AddCapability(new RandomAnimations());
            entity.AddCapability(new LinearMovement());
            entity.AddCapability(new Gravity(gameBounds.H));

            if (isWallCrawler)
            {
                entity.AddCapability(new WallCrawler(gameBounds));
            } else
            {
                entity.AddCapability(new BounceWhenLateralBoundIsHit(0, gameBounds.W));
            }

            return entity;
        }
    }
}
